File: src/main/java/com/meada/whatsapp/profiles/legal/LegalCaseStatus.java
package com.meada.whatsapp.profiles.legal;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

/**
 * Status de um processo (camada 7.2), MATERIALIZADO — espelho 1:1 de
 * {@code frontend/profiles/legal/legal-case-status.ts}. O {@link LegalCaseStatusParityTest}
 * garante a paridade Java↔TS; a CHECK constraint de {@code legal_cases.status} (migration 31)
 * trava os mesmos ids no banco.
 *
 * <p>Diferente do sushi (linear), aqui a transição é LIVRE (qualquer status → qualquer status):
 * o advogado pode reativar um processo arquivado, suspender e reativar, etc.
 */
public enum LegalCaseStatus {
    ATIVO("ativo", "Ativo"),
    SUSPENSO("suspenso", "Suspenso"),
    ARQUIVADO("arquivado", "Arquivado"),
    ENCERRADO("encerrado", "Encerrado");

    private final String id;
    private final String label;

    LegalCaseStatus(String id, String label) {
        this.id = id;
        this.label = label;
    }

    public String id() {
        return id;
    }

    public String label() {
        return label;
    }

    public static Optional<LegalCaseStatus> fromId(String id) {
        if (id == null) {
            return Optional.empty();
        }
        return Arrays.stream(values()).filter(s -> s.id.equals(id)).findFirst();
    }

    public static List<LegalCaseStatus> allActive() {
        return List.of(values());
    }

    /** Transição livre: qualquer status pode ir para qualquer outro (decisão cravada). */
    public boolean canTransitionTo(LegalCaseStatus next) {
        return next != null;
    }

    /**
     * Texto fixo da notificação outbound ao cliente quando o processo entra neste status
     * (decisão 4). null para ATIVO (reativação não notifica). Defensivo juridicamente: sem
     * opinião, sempre orienta a procurar o escritório.
     */
    public String notificationText() {
        return switch (this) {
            case SUSPENSO -> "Informação sobre seu processo: foi colocado em SUSPENSÃO. "
                + "Em caso de dúvida, entre em contato com nosso escritório.";
            case ARQUIVADO -> "Informação sobre seu processo: foi ARQUIVADO. "
                + "Em caso de dúvida, entre em contato com nosso escritório.";
            case ENCERRADO -> "Informação sobre seu processo: foi ENCERRADO. "
                + "Em caso de dúvida, entre em contato com nosso escritório.";
            case ATIVO -> null;
        };
    }
}

File: frontend/profiles/legal/legal-case-status.ts
/**
 * Status de um processo (camada 7.2) — espelho 1:1 de
 * src/main/java/com/meada/whatsapp/profiles/legal/LegalCaseStatus.java.
 *
 * O LegalCaseStatusParityTest (backend) garante que os ids aqui e no enum Java nunca divergem.
 * A CHECK constraint de legal_cases.status (migration 31) trava os mesmos ids no banco.
 * Transição é LIVRE (qualquer status → qualquer status). notificationText: null para 'ativo'.
 */
export const LEGAL_CASE_STATUSES = [
  {
    id: 'ativo',
    label: 'Ativo',
    notificationText: null,
  },
  {
    id: 'suspenso',
    label: 'Suspenso',
    notificationText:
      'Informação sobre seu processo: foi colocado em SUSPENSÃO. Em caso de dúvida, entre em contato com nosso escritório.',
  },
  {
    id: 'arquivado',
    label: 'Arquivado',
    notificationText:
      'Informação sobre seu processo: foi ARQUIVADO. Em caso de dúvida, entre em contato com nosso escritório.',
  },
  {
    id: 'encerrado',
    label: 'Encerrado',
    notificationText:
      'Informação sobre seu processo: foi ENCERRADO. Em caso de dúvida, entre em contato com nosso escritório.',
  },
] as const

export type LegalCaseStatus = (typeof LEGAL_CASE_STATUSES)[number]
export type LegalCaseStatusId = LegalCaseStatus['id']

/** Rótulo pt-BR de um status (fallback: o próprio id). */
export function statusLabel(id: string): string {
  return LEGAL_CASE_STATUSES.find((s) => s.id === id)?.label ?? id
}
